common/wireaddr.c

```diff
@@ -197,4 +323,215 @@ bool parse_wireaddr(const char *arg, struct wireaddr *addr, u16 defport,
 		*err_msg = "Error parsing hostname";
 	tal_free(tmpctx);
 	return res;
+
+}
+
+bool parse_tor_wireaddr(const char *arg,u8 *ip_ld,u16 *port_ld)
+{
+	u16 port;
+	char *ip;
+
+	bool res;
+	tal_t *tmpctx = tal_tmpctx(NULL);
+	res = false;
+	ip = tal_strdup(tmpctx,"127.0.0.1");
+	port = 9050;
+	if (!separate_address_and_port(tmpctx, arg,&ip, &port)) {
+		tal_free(tmpctx);
+		return false;
+	}
+	else
+	{
+		assert(strlen(ip) < 16);
+		memcpy(ip_ld,ip,strlen(ip)+1);
+		(*port_ld) = port;
+		res= true;
+	}
+	tal_free(tmpctx);
+	return res;
+}
+
+
+#define MAX_TOR_COOKIE_LEN 32
+#define MAX_TOR_SERVICE_READBUFFER_LEN 255
+#define MAX_TOR_ONION_V2_ADDR_LEN 16
+#define MAX_TOR_ONION_V3_ADDR_LEN 56
+
+struct tor_service_reaching {
+	struct lightningd *ld;
+	u8 buffer[MAX_TOR_SERVICE_READBUFFER_LEN];
+	char *cookie[MAX_TOR_COOKIE_LEN];
+	u8 *p;
+	bool noauth;
+	size_t hlen;
+};
+
+
+static struct io_plan *io_tor_connect_create_onion_finished(struct io_conn *conn, struct tor_service_reaching *reach)
+{
+    u8 tor_dec_bytes[TOR_V3_ADDRLEN];
+
+	if(reach->hlen ==  MAX_TOR_ONION_V2_ADDR_LEN ) {
+	size_t n = tal_count(reach->ld->wireaddrs);
+	tal_resize(&reach->ld->wireaddrs, n+1);
+	reach->ld->wireaddrs[n].type = ADDR_TYPE_TOR_V2;
+	reach->ld->wireaddrs[n].addrlen = TOR_V2_ADDRLEN;
+	reach->ld->wireaddrs[n].port = reach->ld->portnum;
+	b32_decode((u8 *)tor_dec_bytes,(u8 *)reach->buffer,2);
+	memcpy(&reach->ld->wireaddrs[n].addr,tor_dec_bytes, reach->ld->wireaddrs[n].addrlen);
+
+	}
+	/*on the other hand we can stay connected until ln finish to keep onion alive and then vanish*/
+	//because when we run with Detach flag as we now do every start of LN creates a new addr while the old
+	//stays valid until reboot this might not be desired so we can also drop Detach and use the
+	//read_partial to keep it open until LN drops
+	//FIXME: SAIBATO
+	//return io_read_partial(conn, reach->p, 1 ,&reach->hlen, io_tor_connect_create_onion_finished, reach);
+	return io_close(conn);
+}
+
+
+static struct io_plan *io_tor_connect_after_create_onion(struct io_conn *conn, struct tor_service_reaching *reach)
+{
+
+	reach->p = reach->p+reach->hlen;
+
+ if (!strstr((char *)reach->buffer,"ServiceID=")) {
+	if (reach->hlen == 0) return io_close(conn);
```

Comment:
Should be separate lines:

    if (reach->hlen == 0)
            return io_close(conn);